src/client.rs

```diff
@@ -152,48 +152,116 @@ where
         let mut stream =
             Stream::new(&mut this.io, &mut this.session).set_eof(!this.state.readable());
 
-        #[allow(clippy::match_single_binding)]
-        match this.state {
-            #[cfg(feature = "early-data")]
-            TlsState::EarlyData(ref mut pos, ref mut data) => {
-                use std::io::Write;
+        #[cfg(feature = "early-data")]
+        if let TlsState::EarlyData(pos, data) = &mut this.state {
+            use std::io::Write;
+
+            // write early data
+            if let Some(mut early_data) = stream.session.early_data() {
+                let len = match early_data.write(buf) {
+                    Ok(n) => n,
+                    Err(err) => return Poll::Ready(Err(err)),
+                };
+                if len != 0 {
+                    data.extend_from_slice(&buf[..len]);
+                    return Poll::Ready(Ok(len));
+                }
+            }
+
+            // complete handshake
+            while stream.session.is_handshaking() {
```

Comment:
Now the handshake part is repeated 3 times, I think it can be merge into a function